operators/vision/draw_bounding_box.cc

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Note make sure the input image has the format BGR
+
+#include "draw_bounding_box.hpp"
+#include <algorithm>
+#include <gsl/narrow>
+#include <gsl/span>
+#include <vector>
+
+namespace ort_extensions {
+namespace {
+constexpr uint8_t KColorMap[10][3] = {
+    {255, 0, 0}, {0, 255, 0}, {0, 0, 255}, {255, 255, 0}, {255, 0, 255}, {128, 0, 0}, {0, 128, 0}, {0, 0, 128}, {128, 128, 128}, {0, 0, 0}};
+
+// Used to store image data, width, height, and number of channels
+struct ImageView {
+  uint8_t* data;
+  int64_t height;
+  int64_t width;
+  int64_t channels;
+};
+
+// To represent Boxes, [num of boxes, box_info]
+class RankTwoVector {
+ public:
+  RankTwoVector(const std::vector<int64_t>& shape, const float* data) : shape_(shape), data_(data) {
+  }
+
+  [[nodiscard]] const float* GetFirstDimDataPtr(size_t index) const {
+    return data_ + index * shape_.back();
+  }
+
+  [[nodiscard]] int64_t ShapeAtDim(int64_t dim) const {
+    return shape_[dim];
+  }
+
+ private:
+  gsl::span<const int64_t> shape_;
+  const float* data_;
+};
+
+// Draw a line on the image
+void draw_line_horizon(ImageView& image, int64_t point_start, int64_t line_length, const uint8_t* color,
+                       int64_t thickness) {
+  auto stride = image.width * image.channels;
+  for (int64_t i = 0; i < thickness; ++i) {
+    auto* end = image.data + point_start + i * stride + line_length * image.channels;
+    for (auto* start = image.data + point_start + i * stride; start < end; start += image.channels) {
+      std::copy_n(color, image.channels, start);
+    }
+  }
+}
+
+void draw_line_vertical(ImageView& image, int64_t point_offset, int64_t line_length, const uint8_t* color,
+                        int64_t thickness) {
+  auto* point_start = image.data + point_offset;
+  auto* point_end = point_start + line_length * image.width * image.channels;
+  for (; point_start < point_end; point_start += image.width * image.channels) {
+    for (int64_t i = 0; i < thickness; ++i) {
+      // no boundary check
+      std::copy_n(color, image.channels, point_start + i * image.channels);
+    }
+  }
+}
+
+// Draw a box on the image with given thickness and color
+void draw_box(ImageView& image, const float* box,
+              const uint8_t* color, int64_t thickness) {
+  // -------(1)------
+  // |              |
+  //(2)            (4)
+  // |              |
+  // -------(3)------
+  int64_t x_start = static_cast<int64_t>(box[0]);
+  int64_t y_start = static_cast<int64_t>(box[1]);
+  int64_t x_end = static_cast<int64_t>(box[2]);
+  int64_t y_end = static_cast<int64_t>(box[3]);
+
+  auto half_thickness = (thickness - 1) / 2;
```

Comment:
Why -1?

If thickness is 0 or 1 this gives a value of 0. Is that intended?

Would we avoid potential invalid data access and simplify things if we always drew the lines on the inside of the box vs. trying to center the lines on the box?

Reply:
I would expect 'thickness >=1'.  Not sure how I should hanlde this if thickness==0, to return directly or failed?

The reason of "thickness - 1" :
if we want to draw a point in x=2, and we have thickness=2,  then we can draw  the point at {1,2} or {2,3}.  Not sure which one is the best.
  \|-----------------------------------\|
  \|-==--------------------------------\|
  \|--==-------------------------------\|

I don't think drawing the lines on the inside of the box would simplify this and avoid invalid data access. Because we have  to process the boundary whenever. trying to center the lines on the box will make it looks better instead.

Reply:
thickness == 0 is throw in ctor I would expect. It's invalid usage of the op.

Regarding invalid data access, the std::min/std::max adjustment should prevent that. Sorry - overlooked that part.

I'd favour pushing the box slightly out - so for thickness 2 and a box from 2,2, to 8,8 I'd draw on pixels 1 and 2 for left and top and pixels 8 and 9 for bottom and right. 

Might be clearer to call it `auto offset = thickness / 2;` as well given that value doesn't truly represent half the thickness.

The user has no reference point to compare how we draw the boxes to, so 'looks better' seems like it would be very hard to quantify/prove.

Reply:
>The user has no reference point to compare how we draw the boxes to, so 'looks better' seems like it would be very hard to quantify/prove
hmm, It's ture.